src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,302 @@
+            /// the editor should be saved before then.
+            /// Allowed Transitions: <see cref="NoEditor"/> when editor shutdown is complete.
+            /// </summary>
+            EditorClosing
+        }
+
+        private readonly object _lock = new object();
+        private readonly IProjectThreadingService _threadingService;
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IVsShellUtilitiesHelper _shellHelper;
+        private readonly IExportFactory<IProjectFileModelWatcher> _projectFileWatcherFactory;
+        private readonly IExportFactory<ITextBufferStateListener> _textBufferListenerFactory;
+        private readonly ITextBufferManager _textBufferManager;
+
+        private EditorState _currentState = EditorState.NoEditor;
+        private IVsWindowFrame _windowFrame;
+        private IProjectFileModelWatcher _projectFileModelWatcher;
+        private EditProjectFileVsFrameEvents _frameEventsListener;
+        private ITextBufferStateListener _textBufferStateListener;
+
+        [ImportingConstructor]
+        public EditorStateModel(IProjectThreadingService threadingService,
+            UnconfiguredProject unconfiguredProject,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IVsShellUtilitiesHelper shellHelper,
+            IExportFactory<IProjectFileModelWatcher> projectFileModelWatcherFactory,
+            IExportFactory<ITextBufferStateListener> textBufferListenerFactory,
+            ITextBufferManager textBufferManager)
+        {
+            _threadingService = threadingService;
+            _unconfiguredProject = unconfiguredProject;
+            _serviceProvider = serviceProvider;
+            _shellHelper = shellHelper;
+            _projectFileWatcherFactory = projectFileModelWatcherFactory;
+            _textBufferListenerFactory = textBufferListenerFactory;
+            _textBufferManager = textBufferManager;
+        }
+
+        #region Initialization/Destruction
+
+        /// <summary>
+        /// Called by anything attempting to open a project file editor window, usually by a command. This will show the window frame, creating it if
+        /// has not already been created.
+        /// </summary>
+        public async Task OpenEditorAsync()
+        {
+            await _threadingService.SwitchToUIThread();
+            lock (_lock)
+            {
+                // If the editor is already open, just show it and return
+                if (_currentState != EditorState.NoEditor)
+                {
+                    // If we're initializing, _windowFrame might be null. In that case, when the initialization code
+                    // is done, it'll take care of showing the frame.
+                    _windowFrame?.Show();
+                    return;
+                }
+                _currentState = EditorState.Initializing;
```

Comment:
I'd consider pull out the state changes into a separate method that validates and enforces the valid state changes as called out in the above docs.